app/models/patient.rb

```diff
@@ -270,6 +273,15 @@ def confirm_appointment_after_initial_call
     end
   end
 
+  def create_random_pin
+    return if random_pin
+    25.times do
+      self.random_pin = rand(100000..999999)
+      return if self.valid?
+    end
+    raise "Had trouble generating a random pin - please contact tech support"
```

Comment:
Is there something we can do to be fairly certain that this will never hit a collision? Programmatically generating things, or even serial ids padded with zeroes? (e.g. 0001, 0002, 0003, ....) If we're fine with letters in pins, maybe part of a SHA with a salt?

My concern here is that the failure mode isn't really possible for CMs to resolve on their own, and I want to avoid a system where that's the case. So I'd love to explore another option to generate a random pin here where we're a little more confident that there won't be a collision, even if the possibilities are mega remote.

Reply:
I totally agree with your discomfort, and did some digging before doing this. Believe it or not, this turns out to actually be one of the methods recommended in the mongodb docs.

https://www.mongodb.com/blog/post/generating-globally-unique-identifiers-for-use-with-mongodb

In terms of here, we generally have no more than 5000 patients unarchived at any point in time, and this gives 900,000 possible IDs. The odds of this happening then are (5/900)^25 < 3*10^-56. To put that in comparison, the odds of this are about 1 billion times better than those of winning 6 mega millions jackpots back to back (ballpark).

Reply:
I recognize that the odds are low to zero, but I would want to take the raise out of the control flow here regardless - if we had to move everyone to a single tenant instance, for example, this could potentially be a real pain tounwind.

What's the argument against a scoped serial id (DCAF-000001, DCAF-000002, BAF-000001) in favor of a random scrambler?

Reply:
Serial IDs I think would be too visually confusing for CMs/Clinics since they'll be close to each other. They also run the human risk of being used for authentication, in which case a "guessable" ID would be super dangerous.

I do like the scoping with org name though - I'll work on that right now.

Reply:
Can you expound on what 'used for authentication' here means? I'm a little on the sleep-deprived side so having a tough time the scenario - can you walk me thru it in here?